feat(importDashboard): import api, add config_overwrite field.

michael-hoffman-26 opened the pull request:
<!---
Please write the PR title following the conventions at https://www.conventionalcommits.org/en/v1.0.0/
Example:
fix(dashboard): load charts correctly
-->

### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
#### Description
Adding configuration to the import dashboards api.  
when importing a new dashboards, as a zip file,
the user/client can determine is own overwrite configuration.
witch fields(chart|dashboards ...) he want to overwrite.  
#### Front end changes
* Add to the import modal, a `Advance settings` section 
* Inside it the user can determine which fields to  overwrite
* The user need to type "confirm" in order to load the configuration to the request
### pic of the component
<img width="799" alt="image" src="https://user-images.githubusercontent.com/62829859/156178728-00670085-3412-4fb0-afaa-ce4d9b112cc9.png">

Note other component that using the import modal didn't change 

#### Back end changes
* Add `config_overwrite` to `importDashboardsCommand`. 
* The `config_overwrite` determine which modals to overwrite


### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<!--- Skip this if not applicable -->
#### Before


https://user-images.githubusercontent.com/62829859/156184574-dd77560f-03df-4397-8747-2349f4406ab1.mov



#### After

https://user-images.githubusercontent.com/62829859/156183445-1a005c62-d69b-42c8-a78a-44016810ed1e.mov



### TESTING INSTRUCTIONS
<!--- Required! What steps can be taken to manually verify the changes? -->

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

Commit:
fixing the signature of _import

amitmiran137 commented:
@betodealmeida 
we created this ability to enable users to control which models they wants to overwrite depends on their use case.
for an example at NIelsen we would always like to overwrite Dashboards, Charts and Dataset upon exporting a dashboard.
Could you help us complete this one by reviewing and guiding?|

betodealmeida commented:
Oh, this is awesome! I'll take a look tomorrow morning!

betodealmeida commented:
/testenv up

betodealmeida commented on superset/dashboards/api.py:
```diff
@@ -993,8 +999,17 @@ def import_(self) -> Response:
         )
         overwrite = request.form.get("overwrite") == "true"
 
+        config_overwrite: Dict[str, bool] = (
+            json.loads(request.form["configOverwrite"])
+            if "configOverwrite" in request.form
+            else {}
+        )
```
Maybe we should consolidate these two parameters? Something like:

```suggestion
        config_overwrite: Dict[str, bool] = (
            json.loads(request.form["configOverwrite"])
            if "configOverwrite" in request.form
            else {}
        )
        if request.form.get("overwrite") == "true":
            config_overwrite["dashboard"] = True
```

We'd then pass only `config_overwrite` to `ImportDashboardsCommand`. We could add the same logic to the other import commands, which would save us from having to define a custom `run` method in `superset/dashboards/commands/importers/v1/__init__.py`.

betodealmeida commented on superset/dashboards/api.py:
```diff
@@ -954,6 +955,11 @@ def import_(self) -> Response:
+                    configOverwrite:
+                      description: determine which fields to overwrite
+                      type: Dict<str, bool>
+                      example: {'dashboards': True, 'charts': False ...}
```
Nit, the payload is JSON, not Python:

```suggestion
                      example: {"dashboards": true, "charts": false ...}
```

betodealmeida commented on superset/commands/importers/v1/__init__.py:
```diff
@@ -135,6 +138,8 @@ def _prevent_overwrite_existing_model(  # pylint: disable=invalid-name
         self, exceptions: List[ValidationError]
     ) -> None:
         """check if the object exists and shouldn't be overwritten"""
+        if self.config_overwrite:
+            return
         if not self.overwrite:
```
We can simplify this to:

```suggestion
        if not self.overwrite and not self.config_overwrite:
```

Though maybe we should expand the logic a little bit. If `self.config_overwrite["dashboards'"]` is false, eg, we still want to check the UUIDs to see if the import would result in an overwrite, at least to catch the error early (it would fail later in the stack).

Maybe something like:

```python
        if not self.overwrite and not self.config_overwrite.get(f"{self.model_name}s"):
```

(But then maybe we should use singular form in `config_overwrite`, to align with the `model_name` attribute.)

michael-hoffman-26 replied:
Hi @betodealmeida,  thank you for going over this PR.  
This PR aims to make the `"import modules flow"` more friendly, and configurability.  
While developing this feat I got into complexibility because I didn't want to break the flow others created.  

In the current `"import module flow"`, its works like this.  
1. upload the file.  
2. verify you don't overwrite existing models.  
3. if so, ask for permission, then overwrite the model.      

And about what you mentioned in the parenthesis ("it would fail later in the..") I didn't see it happening when I examen it.   
I comment on the validation part (on master) and upload a dashboards model (which contains dashboards that already exist) and I didn't get any errors.  indeed the import didn't overwrite any dashboard because the default is `overwrite==false`.   

In the  **new** `"import modules flow"` it will work like this.
1. Upload your file.
2. setup your configuration. and upload it.  

According to what's going now on the code,  
I saved backup capability to don't break the flow of validation.  

But if you will agree with me, I think that now, using the new flow we can delete the validate overwriting strategy, from both backend and frontend.

michael-hoffman-26 replied:
So in short I don't see an added value in the `overwrite` param in the `import/module` API,   
after we can configure the overwrite

betodealmeida commented on superset-frontend/src/views/CRUD/dashboard/DashboardList.tsx:
```diff
@@ -67,6 +67,13 @@ const CONFIRM_OVERWRITE_MESSAGE = t(
     'sure you want to overwrite?',
 );
 
+const OVERWRITE_FIELDS: string[] = [
+  'dashboards',
+  'charts',
+  'datasets',
+  'databases',
+];
```
See my comment below, we might want to have these in the singular form to align with the `ImportModelsCommand.model_name` attribute.

michael-hoffman-26 replied:
OK got it, good idea, I will change it to a singular form

betodealmeida commented on superset-frontend/src/components/ImportModal/index.tsx:
```diff
@@ -261,11 +291,96 @@ const ImportModelsModal: FunctionComponent<ImportModelsModalProps> = ({
+      <StyledInputContainer>
+        <div className="control-label">
+          {t(
+            `Type "${expectedInput}" to confirm loading the config, ` +
+              `otherwise the configuration will not loaded`,
```
```suggestion
            'Type "%(expectedInput)s" to confirm loading the config, ' +
              'otherwise the configuration will not loaded', expectedInput
```

michael-hoffman-26 replied:
Done

betodealmeida commented on superset-frontend/src/components/ImportModal/index.tsx:
```diff
@@ -261,11 +291,96 @@ const ImportModelsModal: FunctionComponent<ImportModelsModalProps> = ({
+    Object.values(configOverwriteStatus).some(status => status);
+
+  const confirmOverwriteConfigByTyping = (): JSX.Element => {
+    const expectedInput = `${t('CONFIRM')}`;
```
```suggestion
    const expectedInput = t('CONFIRM');
```

betodealmeida commented on superset-frontend/src/components/ImportModal/index.tsx:
```diff
@@ -261,11 +291,96 @@ const ImportModelsModal: FunctionComponent<ImportModelsModalProps> = ({
+      key={field}
+      checked={configOverwriteStatus[field]}
+    >
+      {`Overwrite ${field}`}
```
```suggestion
      {t('Overwrite %(field)s', field)}
```

michael-hoffman-26 replied:
DOne

michael-hoffman-26 replied:
DOne

betodealmeida commented on superset-frontend/src/components/ImportModal/index.tsx:
```diff
@@ -261,11 +291,96 @@ const ImportModelsModal: FunctionComponent<ImportModelsModalProps> = ({
+        <>
+          <strong>{t('Be careful.')} </strong>
+          {t(
+            'Overwriting might cause you to lose some of your work. Are you' +
```
```suggestion
            'Overwriting might cause you to lose some of your work. Are you ' +
```

betodealmeida reviewed (commented):
This looks great!

A few comments:

1. For the JS translatable strings, I'm not sure if `t()` works with backtick string interpolation. All the use cases I've seen are with the old-school `%s` marker, and that's what I see in the docs. If backtick works correctly with the translation library please ignore my comments (and let me know!).
2. For consistency, we should add the same "advanced" options to the other import dialogs, but it doesn't have to be in this PR. But I would suggest considering consolidating the backend logic at least, so that the API passes the `config_overwrite` dictionary to all the import commands, not just dashboards — ie, `config_overwrite` would replace `overwrite`, and in the other models (chart, eg) it would simply have an entry for `charts` until we update the frontend. I also suggested changing `config_overwrite` to have keys in the singular ("dashboard" instead of "dashboards"), since that would simplify this logic.
3. Can you add some unit and/or integration tests? Just so we know we're not risking overwriting the users assets if there's a bug somewhere.

yousoph commented:
/testenv up FEATURE_VERSIONED_EXPORT = True

yousoph commented:
/testenv up FEATURE_VERSIONED_EXPORT=True

Commit:
fix(frontend react key warnings): update key of SubMenu

Commit:
typo(import_models-config_overwrite): fixing the translation

yousoph commented:
/testenv up FEATURE_VERSIONED_EXPORT=True

yousoph commented:
I like the flexibility in overwriting that this would add, thanks @michael-hoffman-26 !

A few thoughts: 
1. Agreed with @betodealmeida that it would be great to be consistent and have a similar flow for charts and datasets. 
2. Does the "Confirm" on the advanced settings proceed with the import? If so, I'd suggest using text similar to what we currently have on the overwrite prompt in the below screenshot rather than referring to the import file as a config. 
"Overwriting might cause you to lose some of your work. Are you sure you want to overwrite?"
3. I noticed that if I import an existing dashboard without making selections on the advanced options first, I get into this state, which is confusing : 
![image](https://user-images.githubusercontent.com/10627051/159188356-ca07af7d-0fec-4ec0-a33f-f3010ddebe9b.png)
Could we prompt the user to select only after we know that the file contains objects that already exists? This way, users won't see these options if they don't have any overwrites in their import, and we could do away with the double confirmation that shows up in this screenshot 

Flow would be: 
* Select file with no advanced options showing
* User starts import 
* If import doesn't overwrite anything, proceed with the import. 
* If import does overwrite, then show users the fields in advanced options and allow them to select which objects should be overwritten. Once confirmed, import proceeds. 

@jess-dillard - curious to hear your thoughts here as well!

eschutho committed:
chore: update updating with druid no sql deprecation (#19262)

* update updating with druid no sql deprecation

* Update UPDATING.md

Co-authored-by: Beto Dealmeida <roberto@dealmeida.net>

Co-authored-by: Beto Dealmeida <roberto@dealmeida.net>

stephenLYZ committed:
feat: improve color consistency (save all labels) (#19038)

dpgaspar committed:
chore: remove deprecated celery cli (#19273)

* chore: remove deprecated celery cli

* remove configs and UPDATING

prassanna-helixsense-com committed:
Update README.md (#19270)

dpgaspar committed:
chore: remove PUBLIC_ROLE_LIKE_GAMMA deprecated config key (#19274)

michael-hoffman-26 commented:
Hi, @yousoph thanks for going over this PR.  
About this:  
>A few thoughts:

1. Totaly agree, can be done in farther PR.
2. I did write a warning message with `Overwriting warning msg`,  
"Overwriting might cause you to lose ....".
If the user types `CONFIRM` the configuration will be loaded., otherwise it won't.  
3. I see 2 options for the import flow

#### Option one 
Exactly your flow, what that you suggested.  


#### Option two.
- Select file.
- config the overwrite under advance settings.
- import the model, with configuration attached.

This way when uploading we will do exactly what the configuration setup says.  
If we passed false in the configuration,   
1. no overwritten will be done.   
2. we won't tell the user warning about `you are going to overwrite, please confirm overwrite....` because he determine no overwrite will be done.    


in my solution right now I saved a backup capability to the old flow.  
But after we config the overwrite, I don't see why we should keep warning the user if he is about to overwrite.   

So we need to agree on the right flow, and then I will continue.  
I do see a common sense in your flow as well
But I wanted to show our point of view.

Tnx

PApostol committed:
Various docstring fixes (#18221)

eschutho committed:
fix dataset update table (#19269)

suddjian committed:
feat: SupersetClient config to override 401 behavior (#19144)

* wip

* feat: make 401 responses configurable in SupersetClient

* sort

* guest unauthorized handler

* add toast container to embedded app

* add option for toast presenter to go at the top

* remove confusing comms logging

* lint

* Update superset-frontend/src/embedded/index.tsx

* type correction

john-bodley committed:
fix(presto/trino): Add TIME/TIMESTAMP WITH TIME ZONE (#19263)

Co-authored-by: John Bodley <john.bodley@airbnb.com>

villebro committed:
feat(key-value): add superset metastore cache (#19232)

geido committed:
chore: Remove logo forced width (#19049)

* Remove logo forced width

* improve styling

* improve paddings

* Update UPDATING.md

* Fixes a typo

* Add fixed height

Co-authored-by: Michael S. Molina <70410625+michael-s-molina@users.noreply.github.com>
Co-authored-by: Evan Rusackas <evan@preset.io>

jess-dillard commented:
@michael-hoffman-26 I think @yousoph 's flow suggestion is less clicks/overhead for the user – in either situation, I agree we should only be showing 1 "Confirm" field. In this case it would be confirming both the override and granular override suggestions at the same time. 

UI-wise, let's stack the checkboxes vertically instead of horizontally as they will be easier to scan/read that way. Bonus points for a "select all" checkbox at the top of the list that would be on by default :)

pkdotson committed:
feat: allow uploads in crud view (#18953)

* feat: allow uploads in crud view

* fix merge conflict and fix ts

* fix import

* fix tests

* fix lint

* remove unused var

* fix underline flash and alignment

* fix offset

* fix icon alignment

* fix labels and css issues

* make drowdown primary all the time

* make global

* fix lables

* add upload perms to utils

* remove unused code

* add suggested changes

* update menuright

diegomedina248 committed:
chore: add missing keys to components inside lists (#19161)

prosdev0107 committed:
fix(dashboard): make to show the correct owned objects (#19223)

* fix(dashboard): make to show the correct owned objects

* fix(dashboard): make to filter is reusable

* fix(homepage): make sure the type Array<Filters>

diegomedina248 committed:
fix: most Recently Selected Table Should Appear at the Top of the List on the Left Panel (#19258)

suddjian committed:
allow overriding the guest token PyJWT instance (#19293)

suddjian committed:
Update ToastPresenter.tsx (#19296)

kgabryje committed:
chore: Make font-weights themable, fix font faces (#19236)

* fix(fonts): Import all necessary font packages

* Make html tags themable

* Set bold font weight to 600, add medium font weight to theme

* Replace hard coded font weights with theme variables

* Change some font weight light elements to normal

* Fix tests

* Fix bug in pivot table

* Address code review comments

michael-hoffman-26 commented:
Ok, @jess-dillard  @yousoph .  
I will update it to your flow, and the description of this PR

michael-s-molina committed:
fix: Don't allow duplicated tag values in the Select (#19283)

* fix: Don't allow duplicated tag values in the Select

* Addresses comments and adds test

betodealmeida committed:
feat: API for asset sync (#19220)

* feat: API for asset sync

* Add unit tests.

* Improve tests

* Move files

* Add more tests

villebro committed:
fix(cache): only warn about fallback cache for non-debug mode (#19305)

geido committed:
fix: Logo resizing on page load (#19307)

* fix resizing logo

* Refactor

rusackas committed:
chore: fixing a pluralization typo (#19295)

AAfghahi committed:
docusearch (#19289)

zhaoyongjie committed:
feat(advanced analysis): support MultiIndex column in post processing stage (#19116)

villebro committed:
fix(dashboard): fix default filter bar visibility + add docs (#18741)

* fix(dashboard): fix default filter tab visibility + add tests

* fix types

* lint

* rename docs + add double bang to length

dpgaspar committed:
ci: add more code owners to helm chart source (#19322)

dpgaspar committed:
chore: bump FAB to 3.4.5 (#19323)

sujiplr committed:
docs: SECRET_KEY Rotation Documentation (#19233)

* SECRET_KEY Rotation

Additional documentation for SECRET_KEY rotation and SECRET_KEY setting up.

* Bumped the helm chart version to 0.5.11

Bumped the helm chart version for the new changes.

* Removed the default secret key value from the configuration docs.

Removed the default secret key value from the configuration docs.

stephenLYZ committed:
chore: add type checking in package tests directory (#19320)

Commit:
feat(config_overwrite): add ability to config which models to overwrite.

most changes done in the backend.

diegomedina248 committed:
fix: regression on Select component when handling null values (#19326)

prosdev0107 committed:
fix(explore): make to convert null to N/A in view results (#19316)

* fix(explore): make to convert null to N/A in view results

* fix(explore): make to null formatter move before timeFormatter

kgabryje committed:
fix(dashboard): Chart stuck in loading state when when datasets request and chart request fail (#19327)

rusackas commented:
@michael-hoffman-26 (CC @amitmiran137) is there any intent to rebase this PR and see it through? Seems like a valuable addition that got a bit abandoned somewhere along the way.
